### PR TITLE
Add ASTScope support for `@differentiable` attribute.

### PR DESCRIPTION
**Pull request by dan-zheng**
`@differentiable` attribute where clauses may refer to generic
parameters from some generic context.

Without special ASTScope support for `@differentiable` attributes,
`ASTScopeLookup.cpp` logic tries to resolve the generic parameter
DeclNames in the where clause based on source location alone
(`ASTScopeImpl::findChildContaining`) and fails.

The fix is to add a special `DifferentiableAttributeScope`, mimicking
`SpecializeAttributeScope`. Every `@differentiable` attribute has its
own scope, derived from the declaration on which it is declared.
Unlike `@_specialize`, `@differentiable` may also be declared on
`AbstractStorageDecl` declarations (subscripts and variables).

Resolves [TF-815](https://bugs.swift.org/browse/TF-815).

---

`Decl::getSourceRangeIncludingAttrs` should not consider implicit
`@differentiable` attributes generated during `@differentiating`
attribute type-checking.

[TF-835](https://bugs.swift.org/browse/TF-835) tracks robust lowering for `@differentiating` attributes
that does not involve generating implicit `@differentiable` attributes,
circumventing this issue.

---

Unblocks `swift-DEVELOPMENT-SNAPSHOT-2019-09-24-a` merge into `tensorflow`.

### CODE REVIEW

**Comment by dan-zheng** [comment]
cc @davidungar: your review would be appreciated!

`@differentiable` attribute requires special ASTScope support because it may contain a where clause, similar to `@_specialize`:
```swift
@differentiable(where T: Differentiable)
func id<T>(_ x: T) -> T { x }
```

I mostly copied logic from `SpecializeAttributeScope`. One difference is that `@differentiable` attribute may also be declared on `AbstractStorageDecl` declarations - [see tests](https://github.com/apple/swift/pull/27451/files#diff-64661dadce8dd5a75b9173ec8f59dc5fR32).

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/AST/UnqualifiedLookup.cpp [hunk 1]
```diff
@@ -485,12 +485,7 @@ void UnqualifiedLookupFactory::performUnqualifiedLookup() {
       DC, initialIsCascadingUse};
   const bool crosscheckUnqualifiedLookup =
       Ctx.LangOpts.CrosscheckUnqualifiedLookup;
-  // SWIFT_ENABLE_TENSORFLOW
-  // NOTE(TF-815): using AST scopes for lookup causes standard library
-  // type-checking for `@differentiable` attributes to fail.
-  if ((false)) {
-  // if (useASTScopesForLookup()) {
-  // SWIFT_ENABLE_TENSORFLOW END
+  if (useASTScopesForLookup()) {
```
Comment (dan-zheng) [inline lib/AST/UnqualifiedLookup.cpp]: Note: this reverts an [earlier workaround hack](https://github.com/apple/swift/commit/69ab762d68f8fc6564eda3f4390849e492b06e48).

**Comment by dan-zheng** [comment]
There's one remaining test failure, which is known and orthogonal.
```
Failing Tests (1):
    Swift(linux-x86_64) :: AutoDiff/transposing_attr_type_checking.swift

  Expected Passes    : 10936
  Expected Failures  : 25
  Unsupported Tests  : 1366
  Unexpected Failures: 1
```

**Review by rxwei — APPROVED** [review]

**Review by burmako — APPROVED** [review]
Thank you for the detailed comments! From what I can see, this looks reasonable.

**Review by BradLarson — APPROVED** [review]

**Pull request merged**